### ISSUE TITLE
Requests is slow to upload large files

### ISSUE THREAD

**Issue by Tech356** [kennethreitz/requests#2181]
I am using requests to POST data to a server. In testing with a 14.4 MB file, `requests` took 16.5 seconds and `cURL` took 3.3 seconds. The following image shows the 3 tests I ran. The first bump is using `requests` with `MultipartEncoder` from `requests-toolbelt`. The second bump is just using `requests`. The last bump is `cURL`. 
![request_speed](https://cloud.githubusercontent.com/assets/446219/4048836/621e85ea-2d47-11e4-8fbc-1bc5f07e50c3.PNG)

I am using Python 3.4.0, requests 2.3.0, and requests-toolbelt 0.3.1 on Windows 7 x64.

Are there any settings I can tune to speed up `requests`?

**Comment by Lukasa** [comment]
That's very interesting: can I see the code?

**Comment by sigmavirus24** [comment]
So if you're using the `MultipartEncoder` then what happens is that the object is passed down to `httplib` which has a fixed read size of `8192` bytes. So it takes 1,844 reads to upload the file. If you're passing a file-like object to requests the same thing happens.

I assume that the graph you show is of your memory consumption. `cURL` is a bit more advantageous. Unfortunately there's no way for you stream files faster. If your server can handle `chunked` Transfer-Encoding you might be able to pass an iterator that will return larger chunks but I'm not sure if `httplib` handles those and if it does, how it does so.

In short, there are no settings you can use. This is a fundamental (and unchangeable) limitation of `httplib` (or as it is called on 3.4 `http`).

**Comment by Lukasa** [comment]
@sigmavirus24 Nope, that's a graph of network throughput (IIRC). We should be able to get this at least as fast, even if it means working out what the hell httplib is doing.

**Comment by sigmavirus24** [comment]
Here are the relevant sections of [`http`](http://hg.python.org/cpython/file/5c585929658a/Lib/http/client.py#l865) and [`httplib`](http://hg.python.org/cpython/file/37ebb0d44808/Lib/httplib.py#l818)

**Comment by Lukasa** [comment]
Ugh, that's terrible. It really should use `sendfile` if that's available. (Which I imagine is what curl does.)

**Comment by kennethreitz** [comment]
@sigmavirus24  we should document this wealth of information you have in your brain :)

**Comment by kennethreitz** [comment]
We could add a "low-level details" section to the documentation. I think that would be wildly advanteous to our power users.

**Comment by Lukasa** [comment]
That's a neat idea. We can explain some weird edge-cases, like why the read timeout works the way it works.

**Comment by Tech356** [comment]
@sigmavirus24 The graph is from the networking tab of Windows Task Manger. It is indeed showing network throughput. 

Here is the code for `requests` using `MultipartEncoder`. 

```
def upload_artifact_multi(self, group, version, file_path, artifact, packaging):
        head, tail = os.path.split(file_path)

        assert all([self.repository, packaging, group, artifact, version, file_path, tail])
        post_fields = [
            ('hasPom', 'false'),
            ('r', self.repository),
            ('e', packaging),
            ('g', group),
            ('a', artifact),
            ('v', version),
            ('p', packaging),
            ('file', (tail, open(file_path, 'rb'), 'application/octet-stream'))
        ]
        m = MultipartEncoder.from_fields(fields=post_fields)

        url = self._nexus_upload_url()
        r = self.session.post(url, data=m, headers={'Content-Type': m.content_type})
        r.raise_for_status()
        return r
```

And here is the code that is just using `requests`

```
def upload_artifact(self, group, version, file_path, artifact, packaging):
    # only works for files smaller then 2GiB
    if os.path.getsize(file_path) > (2 * 10 ** 9):
        raise NotImplementedError("Cannot handle files larger then 2GiB")

    form_data = {
        'hasPom': 'false',
        'r': self.repository,
        'e': packaging,
        'g': group,
        'a': artifact,
        'v': version,
        'p': packaging,
    }
    url = self._nexus_upload_url()
    r = self.session.post(url, data=form_data, files={'file': open(file_path, 'rb')})
    r.raise_for_status()
    return r
```

And lastly, the command for `cURL`

```
curl -v -F r="repo" -F hasPom=false -F e="ext" -F g="test.speed" -F a="test1" -F v="3.1.1" -F p="ext" -F file=@"file.ext" -u "user":"pass" http://example.com/upload
```

Here is the output from `cURL`

```
* Hostname was NOT found in DNS cache                                                                                 
*   Trying 10.0.0.73...                                                                                           
* Connected to example.com (10.0.0.73) port 80 (#0)                                                          
* Server auth using Basic with user 'user'                                                                        
> POST /upload HTTP/1.1                                                           
> Authorization: Basic d32SrwoxTADXP5vavYm90NQc=                                                                       
> User-Agent: curl/7.37.1                                                                                             
> Host: example.com                                                                                              
> Accept: */*                                                                                                         
> Content-Length: 14737022                                                                                            
> Expect: 100-continue                                                                                                
> Content-Type: multipart/form-data; boundary=------------------------ed123815827ca70f                                
>                                                                                                                     
< HTTP/1.1 100 Continue                                                                                               
< HTTP/1.1 201 Created                                                                                                
< Date: Tue, 26 Aug 2014 17:29:45 GMT                                                                                 
* Server Nexus/2.8.1-01 is not blacklisted                                                                            
< Server: Nexus/2.8.1-01                                                                                              
< X-Frame-Options: SAMEORIGIN                                                                                         
< X-Content-Type-Options: nosniff                                                                                     
< Set-Cookie: rememberMe=deleteMe; Path=/nexus; Max-Age=0; Expires=Mon, 25-Aug-2014 17:29:45 GMT                      
< Content-Type: text/html; charset=UTF-8                                                                              
< Date: Tue, 26 Aug 2014 17:29:48 GMT                                                                                 
* Server Noelios-Restlet-Engine/1.1.6-SONATYPE-5348-V8 is not blacklisted                                             
< Server: Noelios-Restlet-Engine/1.1.6-SONATYPE-5348-V8                                                               
< Content-Length: 86                                                                                                  
<                                                                                                                     
{"groupId":"test.speed","artifactId":"test1","version":"3.1.1","packaging":"ext"}
* Connection #0 to host example.com left intact   
```

It does not seem to make a difference whether `MultipartEncoder` is in use. Both uploads using `requests` take ~16.5 seconds.

**Comment by Lukasa** [comment]
Yeah, I gotta believe that curl is avoiding the small `send` chunks.

**Comment by Lukasa** [comment]
Oh how I hate `httplib`. Another item on my list of things to fix.

**Comment by Lukasa** [comment]
Stupid httplib.

**Comment by lifeofdave** [comment]
Sorry to bring this back from the dead, but have there been any developments on this front?

**Comment by Lukasa** [comment]
There have not. Sorry!

**Comment by sigmavirus24** [comment]
@lifeofdave you could try to adapt something like https://github.com/sigmavirus24/requests-toolbelt/pull/84 to your needs. That code is not merged into the toolbelt and not really well tested. That said, it forces the a larger read than what httplib asks for meaning it will send more data than httplib intends. That said, it's probably not a good long-term solution.

**Comment by kpflugshaupt** [comment]
Hi,

bringing this back from the dead once more -- sorry!

I saw that httplib seems to have added a `blocksize` parameter for Python 3.7 ([Documentation Link](https://docs.python.org/3/library/http.client.html)). It defaults to 8k, but higher values can be supplied.

Looking at the Requests docs, I cannot say if it's possible to set this parameter so it gets passed down to `http.client.HTTPConnection`. Is there a way to achieve this?

Regards
Kaspar

**Comment by sigmavirus24** [comment]
At the moment, no

**Comment by happy-monk** [comment]
@kpflugshaupt I've managed to change it like this:
```python
from http.client import HTTPConnection
HTTPConnection.__init__.__defaults__ = tuple(
    x if x != 8192 else 64*1024
    for x in HTTPConnection.__init__.__defaults__
)
```

**Comment by kpflugshaupt** [comment]
> @kpflugshaupt I've managed to change it like this:
> 
> ```python
> 
> from http.client import HTTPConnection
> 
> HTTPConnection.__init__.__defaults__ = tuple(
> 
>     x if x != 8192 else 64*1024
> 
>     for x in HTTPConnection.__init__.__defaults__
> 
> )
> 
> ```
> 
> 

I will try that out! Thanks for posting.

**Comment by hygohil** [comment]
@kpflugshaupt Did this solution worked?

> > @kpflugshaupt I've managed to change it like this:
> > ```python
> > from http.client import HTTPConnection
> > 
> > HTTPConnection.__init__.__defaults__ = tuple(
> > 
> >     x if x != 8192 else 64*1024
> > 
> >     for x in HTTPConnection.__init__.__defaults__
> > 
> > )
> > ```
> 
> I will try that out! Thanks for posting.

**Comment by kpflugshaupt** [comment]
> @kpflugshaupt Did this solution worked?
> 
> 
> 
> > > @kpflugshaupt I've managed to change it like this:
> 
> > > ```python
> 
> > > from http.client import HTTPConnection
> 
> > > 
> 
> > > HTTPConnection.__init__.__defaults__ = tuple(
> 
> > > 
> 
> > >     x if x != 8192 else 64*1024
> 
> > > 
> 
> > >     for x in HTTPConnection.__init__.__defaults__
> 
> > > 
> 
> > > )
> 
> > > ```
> 
> > 
> 
> > I will try that out! Thanks for posting.
> 
> 
> 
> 

Yes, it works. But I feel it's a bit fragile, as it depends on the current internals of http.client, which may change. 

Still, I may end up using it if I run into performance problems. So far, I avoided them by keeping my transfers small. 

Cheers & thanks
Kaspar
